File: 0-simple-zig/main.js
// Using the browser's built-in WebAssembly API.

const { instance } = await WebAssembly.instantiateStreaming(
  fetch("./simple.wasm"),
);

// Holds the functions marked `export fn`, plus the module's linear memory.
const { add, sum, scratch_ptr, memory } = instance.exports;

// scalar call
const $ = (id) => document.getElementById(id);

$("add").addEventListener("click", () => {
  const result = add(Number($("a").value), Number($("b").value));
  $("addOut").textContent = result;
});

// shared linear memory
// `memory.buffer` is an ArrayBuffer the module and JS both point at.
// Ask the module where its scratch buffer lives, write bytes there, then call sum().
$("sum").addEventListener("click", () => {
  const bytes = $("bytes")
    .value.split(",")
    .map((s) => Number(s.trim()))
    .filter((n) => Number.isFinite(n));

  const ptr = scratch_ptr();
  const view = new Uint8Array(memory.buffer, ptr, bytes.length);
  view.set(bytes);

  $("sumOut").textContent = sum(ptr, bytes.length);
});

File: 0-simple-zig/src/main.zig
// A freestanding WebAssembly module — no std, no runtime, no glue.
// Each `export fn` becomes a function the JavaScript host can call directly.

// Scalar in, scalar out:
export fn add(a: i32, b: i32) i32 {
    return a + b;
}

// JS writes bytes into the module's linear memory, then calls this to sum them:
export fn sum(ptr: [*]const u8, len: usize) u32 {
    var total: u32 = 0;
    for (ptr[0..len]) |b| total += b;
    return total;
}

// Hand the host a pointer into the linear memory it can write into:
var scratch: [256]u8 = undefined;
export fn scratch_ptr() [*]u8 {
    return &scratch;
}
